File: src/main/java/org/javiermf/features/services/rest/ProductsConfigurationFeaturesResource.java
package org.javiermf.features.services.rest;

import org.javiermf.features.services.ProductsConfigurationsService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import javax.ws.rs.*;
import javax.ws.rs.core.Response;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.List;

@Component
@Produces("application/json")
public class ProductsConfigurationFeaturesResource {

    @Autowired
    ProductsConfigurationsService configurationsService;

    @GET
    public List<String> getConfigurationActivedFeatures(@PathParam("productName") String productName,
                                                        @PathParam("configurationName") String configurationName) {
        return configurationsService.getConfigurationActivedFeaturesNames(productName, configurationName);

    }

    @POST
    @Path("/{featureName}")
    public Response addFeatureToConfiguration(@PathParam("productName") String productName,
                                              @PathParam("configurationName") String configurationName,
                                              @PathParam("featureName") String featureName
    ) throws URISyntaxException {
        configurationsService.addFeatureFromConfiguration(productName, configurationName, featureName);
        return Response.created(new URI("/products/" + productName + "/configurations/" + configurationName + "/features/" + featureName)).build();
    }


    @DELETE
    @Path("/{featureName}")
    public Response deleteFeature(@PathParam("productName") String productName,
                                  @PathParam("configurationName") String configurationName,
                                  @PathParam("featureName") String featureName
    ) {
        configurationsService.removeFeatureFromConfiguration(productName, configurationName, featureName);
        return Response.noContent().build();
    }

}

File: src/main/kotlin/org/javiermf/features/services/rest/ProductsConfigurationFeaturesResource.kt
package org.javiermf.features.services.rest

import org.javiermf.features.exceptions.WrongProductConfigurationException
import org.javiermf.features.services.ProductsConfigurationsService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.net.URI
import java.net.URISyntaxException
import javax.ws.rs.*
import javax.ws.rs.core.Response

@Component
@Produces("application/json")
class ProductsConfigurationFeaturesResource {

    @Autowired
    internal var configurationsService: ProductsConfigurationsService? = null

    @GET
    fun getConfigurationActivedFeatures(@PathParam("productName") productName: String,
                                        @PathParam("configurationName") configurationName: String): List<String> {
        return configurationsService!!.getConfigurationActivedFeaturesNames(productName, configurationName)

    }

    @POST
    @Path("/{featureName}")
    @Throws(URISyntaxException::class, WrongProductConfigurationException::class)
    fun addFeatureToConfiguration(@PathParam("productName") productName: String,
                                  @PathParam("configurationName") configurationName: String,
                                  @PathParam("featureName") featureName: String
    ): Response {
        val result = configurationsService!!.addFeatureFromConfiguration(productName, configurationName, featureName)
        return if (result.isValid) {
            Response.created(URI("/products/$productName/configurations/$configurationName/features/$featureName")).build()
        } else {
            throw WrongProductConfigurationException(result.evaluationErrorMessages)
        }

    }


    @DELETE
    @Path("/{featureName}")
    @Throws(WrongProductConfigurationException::class)
    fun deleteFeature(@PathParam("productName") productName: String,
                      @PathParam("configurationName") configurationName: String,
                      @PathParam("featureName") featureName: String
    ): Response {
        val result = configurationsService!!.removeFeatureFromConfiguration(productName, configurationName, featureName)
        return if (result.isValid) {
            Response.noContent().build()
        } else {
            throw WrongProductConfigurationException(result.evaluationErrorMessages)
        }
    }

}
